include/pybind11/cast.h

```diff
@@ -837,6 +864,57 @@ template <typename T> object cast(const T &value,
 template <typename T> T handle::cast() const { return pybind11::cast<T>(*this); }
 template <> inline void handle::cast() const { return; }
 
+template <typename T>
+typename std::enable_if<detail::move_always<T>::value || detail::move_if_unreferenced<T>::value, T>::type move(object &&obj) {
+    if (obj.ref_count() > 1)
+#if defined(NDEBUG)
+        throw cast_error("Unable to cast Python instance to C++ rvalue: instance has multiple references"
+            " (compile in debug mode for details)");
+#else
+        throw cast_error("Unable to move from Python " + (std::string) obj.get_type().str() +
+                " instance to C++ " + type_id<T>() + " instance: instance has multiple references");
+#endif
+
+    typedef detail::type_caster<T> type_caster;
+    type_caster conv;
+    if (!conv.load(obj, true))
+#if defined(NDEBUG)
+        throw cast_error("Unable to cast Python instance to C++ type (compile in debug mode for details)");
+#else
+        throw cast_error("Unable to cast Python instance of type " +
+            (std::string) obj.get_type().str() + " to C++ type '" + type_id<T>() + "''");
+#endif
+
+    // Move into a temporary and return that, because the reference may be a local value of `conv`
+    T ret = std::move(conv.operator T&());
+    return ret;
+}
+
+// Calling cast() on an rvalue calls pybind::cast with the object rvalue, which does:
+// - If we have to move (because T has no copy constructor), do it.  This will fail if the moved
+//   object has multiple references, but trying to copy will fail to compile.
+// - If both movable and copyable, check ref count: if 1, move; otherwise copy
+// - Otherwise (not movable), copy.
+template <typename T> typename std::enable_if<detail::move_always<T>::value, T>::type cast(object &&object) {
+    return move<T>(std::move(object));
+}
+template <typename T> typename std::enable_if<detail::move_if_unreferenced<T>::value, T>::type cast(object &&object) {
+    if (object.ref_count() > 1)
+        return cast<T>(object);
+    else
+        return move<T>(std::move(object));
+}
+template <typename T> typename std::enable_if<detail::move_never<T>::value, T>::type cast(object &&object) {
+    return cast<T>(object);
+}
+
+template <typename T> T object::cast() const & { return pybind11::cast<T>(*this); }
+template <typename T> T object::cast() && { return pybind11::cast<T>(std::move(*this)); }
+template <> inline void object::cast() const & { return; }
```

Comment:
It's a full template specialization, rather than partial, which is legal.  (That's also why it needs the explicit "inline" to avoid a multiple definition link error).